internal/controller/instance_controller.go

```diff
@@ -0,0 +1,1628 @@
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	apimeta "k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8sruntime "k8s.io/apimachinery/pkg/runtime"
+	k8sjson "k8s.io/apimachinery/pkg/runtime/serializer/json"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/utils/ptr"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/cluster"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	mcbuilder "sigs.k8s.io/multicluster-runtime/pkg/builder"
+	mccontext "sigs.k8s.io/multicluster-runtime/pkg/context"
+	mcmanager "sigs.k8s.io/multicluster-runtime/pkg/manager"
+	mcreconcile "sigs.k8s.io/multicluster-runtime/pkg/reconcile"
+
+	"go.datum.net/infra-provider-gcp/internal/config"
+	"go.datum.net/infra-provider-gcp/internal/controller/cloudinit"
+	"go.datum.net/infra-provider-gcp/internal/downstreamclient"
+	datumhandler "go.datum.net/infra-provider-gcp/internal/handler"
+	"go.datum.net/infra-provider-gcp/internal/locationutil"
+	datumsource "go.datum.net/infra-provider-gcp/internal/source"
+	networkingv1alpha "go.datum.net/network-services-operator/api/v1alpha"
+	computev1alpha "go.datum.net/workload-operator/api/v1alpha"
+)
+
+const crossplaneFinalizer = "finalizer.managedresource.crossplane.io"
+
+var errResourceIsDeleting = errors.New("resource is deleting")
+
+var imageMap = map[string]string{
+	"datumcloud/ubuntu-2204-lts":           "projects/ubuntu-os-cloud/global/images/ubuntu-2204-jammy-v20240927",
+	"datumcloud/cos-stable-117-18613-0-79": "projects/cos-cloud/global/images/cos-stable-117-18613-0-79",
+}
+
+var machineTypeMap = map[string]string{
+	"datumcloud/d1-standard-2": "n2-standard-2",
+}
+
+//go:embed cloudinit/populate_secrets.py
+var populateSecretsScript string
+
+// InstanceReconciler reconciles Instances and manages their intended state in
+// GCP
+type InstanceReconciler struct {
+	mgr               mcmanager.Manager
+	Config            config.GCPProvider
+	LocationClassName string
+	DownstreamCluster cluster.Cluster
+}
+
+func (r *InstanceReconciler) Reconcile(ctx context.Context, req mcreconcile.Request) (ctrl.Result, error) {
+	logger := log.FromContext(ctx)
+
+	cl, err := r.mgr.GetCluster(ctx, req.ClusterName)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	ctx = mccontext.WithCluster(ctx, req.ClusterName)
+
+	var instance computev1alpha.Instance
+	if err := cl.GetClient().Get(ctx, req.NamespacedName, &instance); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// Don't do anything if a location isn't set
+	if instance.Spec.Location == nil {
+		return ctrl.Result{}, nil
+	}
```

Comment:
Is this because the instance resource would be created before it's assigned to a location?